bubuku/features/swap_partitions.py

```diff
@@ -0,0 +1,176 @@
+        # find partitions that are candidates to be swapped between "fat" and "slim" brokers
+        swap_partition_candidates = self.__find_all_swap_candidates(self.fat_broker_id, self.slim_broker_id,
+                                                                    topics_stats)
+
+        # smallest partition from slim broker is the one we move to fat broker
+        slim_broker_smallest_partition = min(swap_partition_candidates[self.slim_broker_id], key=attrgetter("size"))
+        if not slim_broker_smallest_partition:
+            _LOG.info("No partitions on slim broker(id: {}) found to swap".format(self.slim_broker_id))
+            return False
+        _LOG.info("Slim broker(id: {}) partition to swap: {}"
+                  .format(self.slim_broker_id, slim_broker_smallest_partition))
+
+        # find the best fitting fat broker partition to move to slim broker
+        # (should be as much as possible closing the gap between brokers)
+        fat_broker_swap_candidates = swap_partition_candidates[self.fat_broker_id]
+        matching_swap_partition = self.__find_best_swap_candidate(fat_broker_swap_candidates, self.gap,
+                                                                  slim_broker_smallest_partition.size)
+
+        # if there is no possible swap that will decrease the gap - just do nothing
+        if not matching_swap_partition:
+            _LOG.info("No candidate from fat broker(id:{}) found to swap".format(self.fat_broker_id))
+            return False
+        _LOG.info("Fat broker(id: {}) partition to swap: {}".format(self.fat_broker_id, matching_swap_partition))
+
+        # write rebalance-json to ZK; Kafka will read it and perform the partitions swap
+        self.to_move = self.__create_swap_partitions_json(slim_broker_smallest_partition, self.slim_broker_id,
+                                                          matching_swap_partition, self.fat_broker_id)
+        scheduled_rebalance = self.__perform_swap(self.to_move)
+        if scheduled_rebalance:
+            _LOG.info("Swap partitions rebalance node was successfully created in ZK")
+        else:
+            _LOG.info("Swap partitions is postponed as there was a rebalance node in ZK")
+        return not scheduled_rebalance
+
+    def __perform_swap(self, swap_json):
+        _LOG.info("Writing rebalance-json to ZK for partitions swap: {}".format(swap_json))
+        return self.zk.reallocate_partitions(swap_json)
+
+    def __find_all_swap_candidates(self, fat_broker_id: str, slim_broker_id: str, topics_stats: dict) -> dict:
+        partition_assignment = self.zk.load_partition_assignment()
+        swap_partition_candidates = {}
+        for topic, partition, replicas in partition_assignment:
+            if topic not in topics_stats or partition not in topics_stats[topic]:
+                continue  # we skip this partition as there is not data size stats for it
+
+            if fat_broker_id in replicas and slim_broker_id in replicas:
```

Comment:
`fat_broker_id` and `slim_broker_id` are strings, `replicas` - array of integers